llvm/lib/Target/AIE/AIE2PreLegalizerCombiner.cpp

```diff
@@ -288,6 +293,89 @@ bool AIE2PreLegalizerCombinerImpl::tryToCombineVectorInserts(
   return true;
 }
 
+/// \returns true if it is possible to combine the below sequence of MIRs
+/// From : %10:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem16.I512),
+/// %2(<32 x s16>), %0(s32), %1(s32)
+///        %20:_(s16) = G_TRUNC %10(s32)
+///        %30:_(s20) = G_SEXT %20(s16)
+/// To :   %10:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem16.I512),
+/// %2(<32 x s16>), %0(s32), %1(s32)
+///        %30:_(s20) = G_TRUNC %10(s32)
+/// Or even:
+/// From : %10:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem8.I512),
+/// %2(<64 x s8>), %0(s32), %1(s32)
+///        %20:_(s8) = G_TRUNC %10(s32)
+///        %30:_(s20) = G_SEXT %20(s8)
+/// To :   %10:_(s32) = G_INTRINSIC intrinsic(@llvm.aie2.vextract.elem8.I512),
+/// %2(<64 x s8>), %0(s32), %1(s32)
+///        %30:_(s20) = G_TRUNC %10(s32)
+/// This also enables S20Narrowing for vextract
+bool AIE2PreLegalizerCombinerImpl::tryToCombineTruncExt(
+    Register DstReg, bool SignVal, unsigned SrcEltSize) const {
+  // Checks if a given register has non-debug user with a specific opcode and
+  // destination size, and return that user.
+  auto GetUseWithOpCode =
+      [](const MachineRegisterInfo &MRI, const Register Reg,
+         const unsigned OpcodeToCheck,
+         const unsigned DstSize) -> std::optional<MachineInstr *> {
+    for (auto &Use : MRI.use_nodbg_instructions(Reg)) {
+      if (Use.getOpcode() == OpcodeToCheck) {
+        const LLT DstRegTy = MRI.getType(Use.getOperand(0).getReg());
+        if (DstRegTy.getSizeInBits() == DstSize)
+          return &Use;
+      }
+    }
+    return std::nullopt;
+  };
+
+  if (auto Trunc =
+          GetUseWithOpCode(MRI, DstReg, TargetOpcode::G_TRUNC, SrcEltSize)) {
+    MachineInstr *TruncMI = Trunc.value();
+    const unsigned ExtOpcode =
+        SignVal ? TargetOpcode::G_SEXT : TargetOpcode::G_ZEXT;
+    const Register UseDstReg = TruncMI->getOperand(0).getReg();
+    // Ensure G_TRUNC has a single non-debug user before safely eliminating it.
+    if (!MRI.hasOneNonDBGUser(UseDstReg))
+      return false;
+    if (auto Ext = GetUseWithOpCode(MRI, UseDstReg, ExtOpcode, 20)) {
+      MachineInstr *ExtMI = Ext.value();
+      MachineIRBuilder MIRBuilder(*ExtMI);
+      MIRBuilder.buildInstr(TargetOpcode::G_TRUNC, {ExtMI->getOperand(0)},
+                            {DstReg});
+      ExtMI->eraseFromParent();
```

Comment:
I suggest to not erase 2 instructions here, because we may corrupt the iterator. Let `ExtMI` to be removed by DCE.

Reply:
Combiner engine may access freed memory:

```
    while (!WorkList.empty()) {
      MachineInstr *CurrInst = WorkList.pop_back_val();
      LLVM_DEBUG(dbgs() << "\nTry combining " << *CurrInst;);
      Changed |= tryCombineAll(*CurrInst);
      WLObserver->reportFullyCreatedInstrs();
    }
```

Reply:
Question is, should we refactor `void makeDeadMI(MachineInstr &MI, MachineRegisterInfo &MRI);`  as a common utility function?

Reply:
We cant erase just the TRUNC as it is input for EXT.

Reply:
You will also erase MI (intrinsic) in the caller function.